### PR TITLE
DEPR: allowing unknown array-likes in merge

### PR DESCRIPTION
**Pull request by jbrockmendel**
Not sure what the current status is w/r/t new deprecations.

Until we make this change, some of our annotations in this module are lies.

### CODE REVIEW

**Comment by jbrockmendel** [comment]
Or we could just change it directly for 2.0.  This is a pretty weird corner case

**Comment by mroeschke** [comment]
Curious if there was a motivating factor from usage in another library or just self discovered?

**Comment by jbrockmendel** [comment]
Self-discovered

File: pandas/core/reshape/merge.py [hunk 1]
```diff
@@ -1139,6 +1155,8 @@ def _get_merge_keys(self):
         # ugh, spaghetti re #733
         if _any(self.left_on) and _any(self.right_on):
             for lk, rk in zip(self.left_on, self.right_on):
+                deprecate_unknown_arraylike(lk)
```
Comment (mroeschke) [inline pandas/core/reshape/merge.py:1158]: Might be a nicer user experience if only 1 warning was raised if `lk` and `rk` were both non-supported array-likes

Comment (jbrockmendel) [reply pandas/core/reshape/merge.py:1158]: isnt that the default behavior of warnings?

Comment (mroeschke) [reply pandas/core/reshape/merge.py:1158]: Ah right.

**Comment by mroeschke** [comment]
> Or we could just change it directly for 2.0. This is a pretty weird corner case

Up to you how strongly you feel about avoiding this in 2.0. Personally, I think we should use the 2.0 hard break for functionality that would be a PITA to deprecate normally.

**Comment by jbrockmendel** [comment]
> Up to you how strongly you feel about avoiding this in 2.0.

Not the end of the world to wait until 3.0

**Comment by mroeschke** [comment]
Cool, then a whatsnew note in 1.6 would probably be good. But hopefully we can do the 1.x deprecation cleaning spree before adding deprecations in 2.0.

File: pandas/tests/reshape/merge/test_merge.py [hunk 2]
```diff
@@ -2720,6 +2722,25 @@ def test_merge_different_index_names():
...
+    left = DataFrame({"A": range(3), "B": range(1, 4)})
+    right = DataFrame({"C": range(2, 5)})
+
+    import dask.array
+
+    arr = dask.array.array([0, 1, 2])  # matches left["A"]
+
+    msg = "will only allow array-like objects that are one of the following types"
+    with tm.assert_produces_warning(FutureWarning, match=msg):
+        res = merge(left, right, left_on="A", right_on=arr)
```
Comment (mroeschke) [inline pandas/tests/reshape/merge/test_merge.py:2738]: Could you also add tests that the FutureWarning shows up when the dask array is passed to `left_on` and `on`?

Comment (jbrockmendel) [reply pandas/tests/reshape/merge/test_merge.py:2738]: huh this now raises for me locally; i wonder if it isnt getting run on the CI?  ive gotta run, will dig into this later

**Comment by jbrockmendel** [comment]
This now raises in Int64Factorizer when I pass in a dask.array (in the test this currently adds).  In 1.5.3 passing a dask array worked for `left_on` and `right_on`, but not `on`; I'm guessing @phofl's Factorizer work changed this.

My initial motivation here was to be able to accurately annotate the merge code, which ATM uses is_array_like but doesn't actually require "ndarray | EA | Series | Index".  Seeing as how the API has already been changed, any objection to making that change explicit+documented by checking for the specific types?

**Comment by phofl** [comment]
I'd be ok with that.

Fixing wouldn't be too hard I guess (we called ensure_...) before, but not sure if it's worth it

**Comment by jbrockmendel** [comment]
Closing, will address in a few months once we confirm no one complains about the changed behavior.

**Pull request closed without merging**